hack/setup-go.sh

```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# script to setup go version with gimme as needed
+# MUST BE RUN FROM THE REPO ROOT DIRECTORY
+
+# read go-version file unless EKSD_GO_IMAGE_TAG & GO_VERSION are set
+GO_VERSION="${GO_VERSION:-"$(cat .go-version)"}"
+EKSD_GO_IMAGE_TAG="${EKSD_GO_IMAGE_TAG:-"${GO_VERSION}"}"
+GO_IMAGE=public.ecr.aws/eks-distro-build-tooling/golang:$EKSD_GO_IMAGE_TAG-gcc
+
+# we don't actually care where the .env files are
+# however, GIMME_SILENT_ENV doesn't trigger re-generating a .env if it
+# already exists and isn't "silent" (no `go version` command in it)
+# so we fix that by changing where the .env is written, ensuring ours
+# is generated from this repo and silent.
+export GIMME_ENV_PREFIX=./bin/.gimme/
+export GIMME_SILENT_ENV=y
+
+# only setup go if we haven't set FORCE_HOST_GO, or `go version` doesn't match
+# go version output looks like:
+# go version go1.14.5 darwin/amd64
+if ! ([ -n "${FORCE_HOST_GO:-}" ] || \
+      (command -v go >/dev/null && [ "$(go version | cut -d' ' -f3)" = "go${GO_VERSION}" ])); then
+    # eval because the output of this is shell to set PATH etc.
+    eval "$(hack/third_party/gimme/gimme "${GO_VERSION}")"
+fi
```

Comment:
Is this to update Go version when building locally? If yes, I would keep the build option same `docker build` because we don't know how Go is being setup locally.

Reply:
Hi @prateekgogia We do keep the same option for `docker build` by: 
```
.image-linux-%:
	docker buildx build --output=type=docker --platform linux/$* \
		--build-arg golang_image=$(shell hack/setup-go.sh) --no-cache \
		--tag $(IMAGE):$(GIT_COMMIT)-linux_$* .
```
Here we explicitly call `setup-go.sh` with build-arg:
```
--build-arg golang_image=$(shell hack/setup-go.sh) --no-cache \
```
In this way, we are able to use exactly the Go image as base image with exactly the same version specified in `.go-version` file

This linked code change is in:
https://github.com/aws/amazon-eks-pod-identity-webhook/pull/206/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R38

Reply:
Basically, this PR will make sure, when we are:
build binaries
do unit tests
build images

We are using the exact specified Go version from `.go-version` in this repo, we call this repo controlled Go version setup.